browser/ui/tabs/tab_tile_model.h

```diff
@@ -0,0 +1,168 @@
+  void TabsAttachedToNewBrowser(TabTileModel* target_data);
+
+ private:
+  friend class TabTileModelBrowserTest;
+
+  FRIEND_TEST_ALL_PREFIXES(TabTileModelBrowserTest,
+                           BreakTile_WithNonExistingTabIsError);
+  FRIEND_TEST_ALL_PREFIXES(TabTileModelBrowserTest,
+                           TileTabs_WithAlreadyTiledTabIsError);
+  FRIEND_TEST_ALL_PREFIXES(TabTileModelBrowserTest, FindTabTile);
+
+  std::vector<TabTile>::iterator FindTabTile(const tabs::TabHandle& tab);
+  std::vector<TabTile>::const_iterator FindTabTile(
+      const tabs::TabHandle& tab) const;
+
+  // When tabs are attached to another browser window and they are tiled tabs,
+  // create tab tiles to that browser by using |tab_tiles|.
+  void Transfer(TabTileModel* other, std::vector<TabTile> tab_tiles);
+
+  void MakeTiledTabsAdjacent(const TabTile& tile, bool move_right_tab = true);
+  bool SynchronizeGroupedState(const TabTile& tile,
+                               const tabs::TabHandle& source,
+                               std::optional<tab_groups::TabGroupId> group);
+  bool SynchronizePinnedState(const TabTile& tile,
+                              const tabs::TabHandle& source);
+
+  void TabDragEnded();
+  void OnTabInserted(const TabStripModelChange::Insert* insert);
+  void OnTabMoved(const TabStripModelChange::Move* move);
+  void OnTabRemoved(const TabStripModelChange::Remove* remove);
+
+  // TabStripModelObserver:
+  void OnTabStripModelChanged(
+      TabStripModel* tab_strip_model,
+      const TabStripModelChange& change,
+      const TabStripSelectionChange& selection) override;
+  void OnTabWillBeRemoved(content::WebContents* contents, int index) override;
+  void TabPinnedStateChanged(TabStripModel* tab_strip_model,
+                             content::WebContents* contents,
+                             int index) override;
+  void TabGroupedStateChanged(TabStripModel* tab_strip_model,
+                              std::optional<tab_groups::TabGroupId> old_group,
+                              std::optional<tab_groups::TabGroupId> new_group,
+                              tabs::TabInterface* tab,
+                              int index) override;
+
+  // Filled up in OnTabWillBeRemoved() and revisited from OnTabRemoved().
+  // These are in pending state until we can decide what to do with them in
+  // OnTabRemoved() by looking into the reason.
+  std::vector<std::pair<content::WebContents*, content::WebContents*>>
+      tiled_tabs_scheduled_to_be_removed_;
+
+  raw_ref<TabStripModel> model_;
+
+  bool is_in_tab_dragging_ = false;
+
+  bool is_in_synch_grouped_state_ = false;
+
+  std::vector<TabTile> tab_tiles_;
+  std::vector<TabTile> tab_tiles_to_be_attached_to_new_window_;
+
+  // As UI is likely to read more frequently than insert or delete, we cache
+  // index for faster look up.
+  base::flat_map<tabs::TabHandle, size_t> tab_tile_index_for_tab_;
```

Comment:
Restored hidden @bridiver 's previous comment.
```
I'm confused about what exactly this is a cache of, what would be the alternative way to look this up and what is the index?
```

This is for finding `TabTile` of `TabHandle` from `tab_tiles` quickly.
index here is the `tab_tiles` vector index.
With this `TabHandle` and index map, we can get `TabTile` that includes `TabHandle` from `tab_tiles_` quickly. W/o this, we have to iterate all tiles in `tab_tiles_` whenever we want to check current tab is in tile or not. or to get tab's tab tile.